src/lib/snark_work_lib/metrics.ml

```diff
@@ -0,0 +1,180 @@
+open Core_kernel
+
+module Time_span_with_json = struct
+  type t = Time.Span.t
+
+  let to_yojson total = `String (Time.Span.to_string_hum total)
+
+  let of_yojson = function
+    | `String time ->
+        Ok (Time.Span.of_string time)
+    | _ ->
+        Error "Snark_worker.Functor: Could not parse timespan"
+end
+
+(*FIX: register_event fails when adding base types to the constructors*)
+module String_with_json = struct
+  type t = string
+
+  let to_yojson s = `String s
+
+  let of_yojson = function
+    | `String s ->
+        Ok s
+    | _ ->
+        Error "Snark_worker.Functor: Could not parse string"
+end
+
+module Int_with_json = struct
+  type t = int
+
+  let to_yojson s = `Int s
+
+  let of_yojson = function
+    | `Int s ->
+        Ok s
+    | _ ->
+        Error "Snark_worker.Functor: Could not parse int"
+end
+
+type Structured_log_events.t +=
+  | Merge_snark_generated of { elapsed : Time_span_with_json.t }
+  [@@deriving register_event { msg = "Merge SNARK generated in $elapsed" }]
+
+type Structured_log_events.t +=
+  | Base_snark_generated of
+      { elapsed : Time_span_with_json.t
+      ; transaction_type :
+          [ `Zkapp_command | `Signed_command | `Coinbase | `Fee_transfer ]
+      ; zkapp_command_count : Int_with_json.t
+      ; proof_zkapp_command_count : Int_with_json.t
+      }
+  [@@deriving
+    register_event
+      { msg =
+          "Base SNARK generated in $elapsed for $transaction_type transaction \
+           with $zkapp_command_count zkapp_command and \
+           $proof_zkapp_command_count proof zkapp_command"
+      }]
+
+type Structured_log_events.t +=
+  | Sub_zkapp_snark_generated of
+      { elapsed : Time_span_with_json.t; kind : [ `Merge | `Segment ] }
+  [@@deriving
+    register_event
+      { msg = "Sub-zkApp SNARK generated in $elapsed of type $kind" }]
+
+let emit_single_metrics ~(single_spec : _ Single_spec.Poly.t)
+    ~data:
+      ({ data = elapsed; _ } :
+        (Core.Time.Stable.Span.V1.t, _) Proof_carrying_data.t ) =
+  match single_spec with
+  | Merge (_, _, _) ->
+      Perf_histograms.add_span ~name:"snark_worker_merge_time" elapsed ;
+      Mina_metrics.(
+        Cryptography.Snark_work_histogram.observe
+          Cryptography.snark_work_merge_time_sec (Time.Span.to_sec elapsed)) ;
+      Merge_snark_generated { elapsed }
+  | Transition (_, ({ transaction; _ } : Transaction_witness.Stable.Latest.t))
+    ->
+      Perf_histograms.add_span ~name:"snark_worker_transition_time" elapsed ;
+      let transaction_type, zkapp_command_count, proof_zkapp_command_count =
+        match transaction with
+        | Mina_transaction.Transaction.Command
+            (Mina_base.User_command.Zkapp_command zkapp_command) ->
+            (* WARN: now this is dead code, if we're using new protocol between
```

Comment:
Let's remove dead code then :)

_It seems like my some of my comments below will become irrelevant_ 😆

Reply:
I was wrong, this is actually used in `combining_result`